### PR TITLE
Add doc for tests-with-coverage quickstart

### PR DESCRIPTION
**Pull request by fabricepipart**
This is the follow up PR of the [PR done on the quickstarts](https://github.com/quarkusio/quarkus-quickstarts/pull/209).

The goal of this PR is to give an example showing how to obtain the coverage of Unit Tests and Integration Tests. Without proper attention, Jacoco fails to analyze the coverage because Quarkus does some bytecode manipulations (for dependencies injection and PanacheEntity if I understood properly).

Example covers
* Measuring the coverage of your Unit Tests
* Measuring the coverage of your Integration Tests
* Separating the execution of your Unit Tests and Integration Tests
* Consolidating the coverage for all your tests

### CODE REVIEW

File: docs/src/main/asciidoc/tests-with-coverage-guide.adoc [hunk 1]
```diff
@@ -0,0 +1,701 @@
...
+                        <phase>integration-test</phase>
+                        <goals>
+                            <goal>test</goal>
+                        </goals>
+                        <configuration>
+                            <excludedGroups>!integration</excludedGroups>
+                            <groups>integration</groups>
+                        </configuration>
+                    </execution>
+                </executions>
```
Comment (aloubyansky) [inline docs/src/main/asciidoc/tests-with-coverage-guide.adoc]: With your config the integration tests will be run twice: as part of the default-test and then as integration tests. Shouldn't it be something like
````
                <executions>
                    <execution>
                        <id>default-test</id>
                        <phase>test</phase>
                        <goals>
                            <goal>test</goal>
                        </goals>
                        <configuration>
                            <excludedGroups>integration</excludedGroups>
                        </configuration>
                    </execution>
                    <execution>
                        <id>integration-tests</id>
                        <phase>integration-test</phase>
                        <goals>
                            <goal>test</goal>
                        </goals>
                        <configuration>
                            <groups>integration</groups>
                        </configuration>
                    </execution>
                </executions>
````

File: docs/src/main/asciidoc/tests-with-coverage-guide.adoc [hunk 2]
```diff
@@ -0,0 +1,701 @@
...
+    </build>
+----
+
+== The coverage does not seem to correspond to the reality
+
+You can now run the tests: `mvn clean verify` As explained earlier, it will run the unit tests first, then the integration tests. And finally, it will generate two separate reports. First a report of the coverage of the unit tests in `target/site/jacoco-ut` then a report of the coverage of the integration tests in `target/site/jacoco-it`.
+
+Given the content of `GreetingResourceTest`, `GreetingResource` should have been covered. But when we open the report `target/site/jacoco-it/index.html`, the class `GreetingResource` is reported with 0% of coverage. But the fact that `GreetingService` is reported as covered shows that the test execution was actually recorded. How come?
+
+During the execution of the generation of the report, you may have noticed a warning:
```
Comment (aloubyansky) [inline docs/src/main/asciidoc/tests-with-coverage-guide.adoc]: During the report generation, you may have...

File: docs/src/main/asciidoc/tests-with-coverage-guide.adoc [hunk 3]
```diff
@@ -0,0 +1,701 @@
...
+    @Inject
+    public GreetingResource(GreetingService service) {
+        this.service = service;
+    }
+...
+}
+----
+
+Some might say that this approach is preferable since the field can be *final* like this. Anyway, in some cases you might need an explicit constructor. And, in that case, the coverage is not reported properly by JaCoCo.
+This is because Quarkus generates a constructor without any parameter and does some bycode manipulations in order to add it to the class. It is what happened here, just before the execution of the integration tests:
```
Comment (aloubyansky) [inline docs/src/main/asciidoc/tests-with-coverage-guide.adoc]: That is what happened here...

File: docs/src/main/asciidoc/tests-with-coverage-guide.adoc [hunk 4]
```diff
@@ -0,0 +1,701 @@
...
+[INFO] --- quarkus-maven-plugin:0.16.0:build (default) @ getting-started-testing ---
+[INFO] [io.quarkus.deployment.QuarkusAugmentor] Beginning quarkus augmentation
+...
+----
+
+As a consequence, JaCoCo does not recognize the classes when it wants to create its report. But wait... there is a solution.
+
+== Instrumenting the classes instead
+
+JaCoCo has two modes. The first one is based on an agent and instruments classes on-the-fly. Unfortunately, this is incompatible with the dynamic classfile transformations that Quarkus does. But there is a second mode called link:https://www.eclemma.org/jacoco/trunk/doc/offline.html[offline instrumentation]. Classes are pre-instrumented in advance via the *jacoco:instrument* Maven goal and during their usage (when the tests are ran), *jacocoagent.jar* must be added in the classpath.
```
Comment (aloubyansky) [inline docs/src/main/asciidoc/tests-with-coverage-guide.adoc]: "But there is a second mode" -> "The second mode is"

Comment (aloubyansky) [reply docs/src/main/asciidoc/tests-with-coverage-guide.adoc]: "must be added in the classpath." - > "must be added to the classpath."

**Review by aloubyansky — CHANGES_REQUESTED** [review]

File: docs/src/main/asciidoc/tests-with-coverage-guide.adoc [hunk 5]
```diff
@@ -0,0 +1,701 @@
...
+
+== Solution
+
+We recommend that you follow the instructions in the next sections and create the application step by step. However, you can go right to the completed example.
+Clone the Git repository: `git clone {quickstarts-clone-url}`, or download an {quickstarts-archive-url}[archive].
+The solution is located in the `tests-with-coverage` directory.
+
+== Starting from a simple project and two tests
+
+Let's start from an empty application that we'll create thanks to the Quarkus maven plugin:
```
[the remark this answers is not in the document]
Comment (fabricepipart) [reply docs/src/main/asciidoc/tests-with-coverage-guide.adoc]: Sure

File: docs/src/main/asciidoc/tests-with-coverage-guide.adoc [hunk 6]
```diff
@@ -0,0 +1,701 @@
...
+== Starting from a simple project and two tests
+
+Let's start from an empty application that we'll create thanks to the Quarkus maven plugin: 
+
+[source,shell,subs=attributes+]
+----
+mvn io.quarkus:quarkus-maven-plugin:create -DprojectGroupId=org.acme -DprojectArtifactId=tests-with-coverage
+----
+
+We'll now, progressively add all the elements necessary to have an application that is properly covered with tests.
```
[the remark this answers is not in the document]
Comment (fabricepipart) [reply docs/src/main/asciidoc/tests-with-coverage-guide.adoc]: Changed

File: docs/src/main/asciidoc/tests-with-coverage-guide.adoc [hunk 7]
```diff
@@ -0,0 +1,701 @@
...
+
+    @Test
+    public void testGreetingService() {
+        GreetingService service = new GreetingService();
+        Assertions.assertEquals("hello Quarkus", service.greeting("Quarkus"));
+    }
+}
+----
+
+But also a `@QuarkusTest`, `GreetingResourceTest.java`:
```
[the remark this answers is not in the document]
Comment (fabricepipart) [reply docs/src/main/asciidoc/tests-with-coverage-guide.adoc]: Agreed. I adjusted both.

File: docs/src/main/asciidoc/tests-with-coverage-guide.adoc [hunk 8]
```diff
@@ -0,0 +1,701 @@
...
+          .pathParam("name", uuid)
+          .when().get("/hello/greeting/{name}")
+          .then()
+            .statusCode(200)
+            .body(is("hello " + uuid));
+    }
+}
+----
+
+One will be our example of a Unit Test and one will be our example of Integration Test.
```
[the remark this answers is not in the document]
Comment (fabricepipart) [reply docs/src/main/asciidoc/tests-with-coverage-guide.adoc]: Modified

File: docs/src/main/asciidoc/tests-with-coverage-guide.adoc [hunk 9]
```diff
@@ -0,0 +1,701 @@
...
+
+== Separating executions for Unit Tests and Integration Tests
```
[the remark this answers is not in the document]
Comment (fabricepipart) [reply docs/src/main/asciidoc/tests-with-coverage-guide.adoc]: Changed

File: docs/src/main/asciidoc/tests-with-coverage-guide.adoc [hunk 10]
```diff
@@ -0,0 +1,701 @@
...
+                    </execution>
+                </executions>
+            </plugin>
+        </plugins>
+    </build>
+    ...
+</project>
+----
+
+NOTE: This way, the `QuarkusTest` instances will are ran during the `integration-test` maven phase while the other JUnit tests will still be ran during the `test` phase.
```
[the remark this answers is not in the document]
Comment (fabricepipart) [reply docs/src/main/asciidoc/tests-with-coverage-guide.adoc]: Changed

File: docs/src/main/asciidoc/tests-with-coverage-guide.adoc [hunk 11]
```diff
@@ -0,0 +1,701 @@
...
+You can all the tests via the command `mvn clean verify` (and you will notice that the two tests are ran at different times).
```
[the remark this answers is not in the document]
Comment (fabricepipart) [reply docs/src/main/asciidoc/tests-with-coverage-guide.adoc]: Changed

File: docs/src/main/asciidoc/tests-with-coverage-guide.adoc (hunk 1 above)
Comment (fabricepipart) [reply docs/src/main/asciidoc/tests-with-coverage-guide.adoc]: I may have missed something but it should not. By default, the tag "integration" is excluded. That is what applies during the test phase. Then during the integration-test phase, the configuration of integration-tests prevails and only ITs are ran.

Here are the logs of the run:
```bash
╰─ mvn clean verify
[INFO] Scanning for projects...
[INFO] 
[INFO] --------------------< org.acme:tests-with-coverage >--------------------
[INFO] Building tests-with-coverage 1.0-SNAPSHOT
[INFO] --------------------------------[ jar ]---------------------------------
[INFO] 
[INFO] --- maven-clean-plugin:2.5:clean (default-clean) @ tests-with-coverage ---
[INFO] Deleting /Users/legone/Documents/git_clones/quarkus-quickstarts/tests-with-coverage/target
[INFO] 
[INFO] --- maven-resources-plugin:2.6:resources (default-resources) @ tests-with-coverage ---
[INFO] Using 'UTF-8' encoding to copy filtered resources.
[INFO] skip non existing resourceDirectory /Users/legone/Documents/git_clones/quarkus-quickstarts/tests-with-coverage/src/main/resources
[INFO] 
[INFO] --- maven-compiler-plugin:3.1:compile (default-compile) @ tests-with-coverage ---
[INFO] Changes detected - recompiling the module!
[INFO] Compiling 2 source files to /Users/legone/Documents/git_clones/quarkus-quickstarts/tests-with-coverage/target/classes
[INFO] 
[INFO] --- jacoco-maven-plugin:0.8.4:instrument (instrument-ut) @ tests-with-coverage ---
[INFO] 
[INFO] --- maven-resources-plugin:2.6:testResources (default-testResources) @ tests-with-coverage ---
[INFO] Using 'UTF-8' encoding to copy filtered resources.
[INFO] skip non existing resourceDirectory /Users/legone/Documents/git_clones/quarkus-quickstarts/tests-with-coverage/src/test/resources
[INFO] 
[INFO] --- maven-compiler-plugin:3.1:testCompile (default-testCompile) @ tests-with-coverage ---
[INFO] Changes detected - recompiling the module!
[INFO] Compiling 2 source files to /Users/legone/Documents/git_clones/quarkus-quickstarts/tests-with-coverage/target/test-classes
[INFO] 
[INFO] --- maven-surefire-plugin:2.22.1:test (default-test) @ tests-with-coverage ---
[INFO] 
[INFO] -------------------------------------------------------
[INFO]  T E S T S
[INFO] -------------------------------------------------------
[INFO] Running org.acme.quickstart.GreetingServiceTest
[INFO] Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.055 s - in org.acme.quickstart.GreetingServiceTest
[INFO] 
[INFO] Results:
[INFO] 
[INFO] Tests run: 1, Failures: 0, Errors: 0, Skipped: 0
[INFO] 
[INFO] 
[INFO] --- jacoco-maven-plugin:0.8.4:restore-instrumented-classes (restore-ut) @ tests-with-coverage ---
[INFO] 
[INFO] --- maven-jar-plugin:2.4:jar (default-jar) @ tests-with-coverage ---
[INFO] Building jar: /Users/legone/Documents/git_clones/quarkus-quickstarts/tests-with-coverage/target/tests-with-coverage-1.0-SNAPSHOT.jar
[INFO] 
[INFO] --- quarkus-maven-plugin:0.19.1:build (default) @ tests-with-coverage ---
[INFO] [io.quarkus.deployment.QuarkusAugmentor] Beginning quarkus augmentation
[INFO] [org.jboss.threads] JBoss Threads version 3.0.0.Beta4
[INFO] [io.quarkus.deployment.QuarkusAugmentor] Quarkus augmentation completed in 1026ms
[INFO] [io.quarkus.creator.phase.runnerjar.RunnerJarPhase] Building jar: /Users/legone/Documents/git_clones/quarkus-quickstarts/tests-with-coverage/target/tests-with-coverage-1.0-SNAPSHOT-runner.jar
[INFO] 
[INFO] --- jacoco-maven-plugin:0.8.4:instrument (instrument-it) @ tests-with-coverage ---
[INFO] 
[INFO] --- maven-surefire-plugin:2.22.1:test (integration-tests) @ tests-with-coverage ---
[INFO] 
[INFO] -------------------------------------------------------
[INFO]  T E S T S
[INFO] -------------------------------------------------------
[INFO] Running org.acme.quickstart.GreetingResourceTest
2019-07-22 19:18:31,260 INFO  [io.qua.dep.QuarkusAugmentor] (main) Beginning quarkus augmentation
2019-07-22 19:18:32,041 INFO  [io.qua.dep.QuarkusAugmentor] (main) Quarkus augmentation completed in 781ms
2019-07-22 19:18:32,487 INFO  [io.quarkus] (main) Quarkus 0.19.1 started in 0.437s. Listening on: http://[::]:8081
2019-07-22 19:18:32,487 INFO  [io.quarkus] (main) Installed features: [cdi, resteasy]
[INFO] Tests run: 2, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 3.146 s - in org.acme.quickstart.GreetingResourceTest
2019-07-22 19:18:33,479 INFO  [io.quarkus] (main) Quarkus stopped in 0.009s
[INFO] 
[INFO] Results:
[INFO] 
[INFO] Tests run: 2, Failures: 0, Errors: 0, Skipped: 0
[INFO] 
[INFO] 
[INFO] --- jacoco-maven-plugin:0.8.4:restore-instrumented-classes (restore-it) @ tests-with-coverage ---
[INFO] 
[INFO] --- jacoco-maven-plugin:0.8.4:report (report-it) @ tests-with-coverage ---
[INFO] Loading execution data file /Users/legone/Documents/git_clones/quarkus-quickstarts/tests-with-coverage/target/jacoco-it.exec
[INFO] Analyzed bundle 'tests-with-coverage' with 2 classes
[INFO] 
[INFO] --- jacoco-maven-plugin:0.8.4:report (report-ut) @ tests-with-coverage ---
[INFO] Loading execution data file /Users/legone/Documents/git_clones/quarkus-quickstarts/tests-with-coverage/target/jacoco-ut.exec
[INFO] Analyzed bundle 'tests-with-coverage' with 2 classes
[INFO] 
[INFO] --- jacoco-maven-plugin:0.8.4:merge (merge-results) @ tests-with-coverage ---
[INFO] Loading execution data file /Users/legone/Documents/git_clones/quarkus-quickstarts/tests-with-coverage/target/jacoco-ut.exec
[INFO] Loading execution data file /Users/legone/Documents/git_clones/quarkus-quickstarts/tests-with-coverage/target/jacoco-it.exec
[INFO] Writing merged execution data to /Users/legone/Documents/git_clones/quarkus-quickstarts/tests-with-coverage/target/jacoco.exec
[INFO] 
[INFO] --- jacoco-maven-plugin:0.8.4:report (post-merge-report) @ tests-with-coverage ---
[INFO] Loading execution data file /Users/legone/Documents/git_clones/quarkus-quickstarts/tests-with-coverage/target/jacoco.exec
[INFO] Analyzed bundle 'tests-with-coverage' with 2 classes
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  11.081 s
[INFO] Finished at: 2019-07-22T19:18:34+02:00
[INFO] ------------------------------------------------------------------------
```

File: docs/src/main/asciidoc/tests-with-coverage-guide.adoc (hunk 2 above)
Comment (fabricepipart) [reply docs/src/main/asciidoc/tests-with-coverage-guide.adoc]: Changed

File: docs/src/main/asciidoc/tests-with-coverage-guide.adoc (hunk 3 above)
Comment (fabricepipart) [reply docs/src/main/asciidoc/tests-with-coverage-guide.adoc]: Changed

File: docs/src/main/asciidoc/tests-with-coverage-guide.adoc (hunk 4 above)
Comment (fabricepipart) [reply docs/src/main/asciidoc/tests-with-coverage-guide.adoc]: Changed

**Comment by fabricepipart** [comment]
Thanks for the review! I applied all the proposed changes. Excepted the one on Surefire configuration. I thought it was clearer like that but if you prefer I can try to explicitly override the default execution done during the test phase.

**Comment by aloubyansky** [comment]
Ok, perhaps I missed something. Your version of the config looks shorter, so if it does work then let's keep it. I'll test it tomorrow and let you know. Thanks!

File: docs/src/main/asciidoc/tests-with-coverage-guide.adoc (hunk 1 above)
Comment (aloubyansky) [reply docs/src/main/asciidoc/tests-with-coverage-guide.adoc]: Looks like you forgot to push your changes. Here is what I see with your config
````
[INFO] --- maven-surefire-plugin:3.0.0-M2:test (default-test) @ tests-with-coverage ---
[INFO] 
[INFO] -------------------------------------------------------
[INFO]  T E S T S
[INFO] -------------------------------------------------------
[INFO] Running org.acme.quickstart.GreetingResourceTest
2019-07-23 11:18:30,524 INFO  [io.qua.dep.QuarkusAugmentor] (main) Beginning quarkus augmentation
2019-07-23 11:18:31,287 INFO  [io.qua.dep.QuarkusAugmentor] (main) Quarkus augmentation completed in 763ms
2019-07-23 11:18:31,593 INFO  [io.quarkus] (main) Quarkus 0.19.1 started in 0.299s. Listening on: http://[::]:8081
2019-07-23 11:18:31,593 INFO  [io.quarkus] (main) Installed features: [cdi, resteasy]
[INFO] Tests run: 2, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 2.678 s - in org.acme.quickstart.GreetingResourceTest
[INFO] Running org.acme.quickstart.GreetingServiceTest
[INFO] Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.001 s - in org.acme.quickstart.GreetingServiceTest
2019-07-23 11:18:32,539 INFO  [io.quarkus] (main) Quarkus stopped in 0.009s
[INFO] 
[INFO] Results:
[INFO] 
[INFO] Tests run: 3, Failures: 0, Errors: 0, Skipped: 0
[INFO] 
[INFO] 
[INFO] --- maven-jar-plugin:2.4:jar (default-jar) @ tests-with-coverage ---
[INFO] Building jar: /home/aloubyansky/playground/new/target/tests-with-coverage-1.0-SNAPSHOT.jar
[INFO] 
[INFO] --- quarkus-maven-plugin:0.19.1:build (default) @ tests-with-coverage ---
[INFO] [io.quarkus.deployment.QuarkusAugmentor] Beginning quarkus augmentation
[INFO] [org.jboss.threads] JBoss Threads version 3.0.0.Beta4
[INFO] [io.quarkus.deployment.QuarkusAugmentor] Quarkus augmentation completed in 974ms
[INFO] [io.quarkus.creator.phase.runnerjar.RunnerJarPhase] Building jar: /home/aloubyansky/playground/new/target/tests-with-coverage-1.0-SNAPSHOT-runner.jar
[INFO] 
[INFO] --- maven-surefire-plugin:3.0.0-M2:test (integration-tests) @ tests-with-coverage ---
[INFO] 
[INFO] -------------------------------------------------------
[INFO]  T E S T S
[INFO] -------------------------------------------------------
[INFO] Running org.acme.quickstart.GreetingResourceTest
2019-07-23 11:18:35,527 INFO  [io.qua.dep.QuarkusAugmentor] (main) Beginning quarkus augmentation
2019-07-23 11:18:36,270 INFO  [io.qua.dep.QuarkusAugmentor] (main) Quarkus augmentation completed in 743ms
2019-07-23 11:18:36,554 INFO  [io.quarkus] (main) Quarkus 0.19.1 started in 0.273s. Listening on: http://[::]:8081
2019-07-23 11:18:36,554 INFO  [io.quarkus] (main) Installed features: [cdi, resteasy]
[INFO] Tests run: 2, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 2.066 s - in org.acme.quickstart.GreetingResourceTest
2019-07-23 11:18:37,516 INFO  [io.quarkus] (main) Quarkus stopped in 0.004s
[INFO] 
[INFO] Results:
[INFO] 
[INFO] Tests run: 2, Failures: 0, Errors: 0, Skipped: 0
````

Comment (aloubyansky) [reply docs/src/main/asciidoc/tests-with-coverage-guide.adoc]: Hold on, perhaps, I misspelled something. As now I re-did it and it works.

Comment (aloubyansky) [reply docs/src/main/asciidoc/tests-with-coverage-guide.adoc]: Ok, I'm resolving this one.

**Comment by aloubyansky** [comment]
Please make sure you pushed your updates. Thanks!

**Comment by fabricepipart** [comment]
It true that *actually* pushing the changes could be useful ... Sorry for that

**Comment by aloubyansky** [comment]
I haven't pushed them yet, have you?

**Comment by fabricepipart** [comment]
... I pushed to the wrong repo. Should be OK this time... Sorry again

**Review by aloubyansky — APPROVED** [review]

**Pull request merged**